spikes/lk_auto_charge/SealStrategy.cpp

```diff
@@ -0,0 +1,112 @@
+// Leka - LekaOS
+// Copyright 2023 APF France handicap
+// SPDX-License-Identifier: Apache-2.0
+
+#include "SealStrategy.h"
+
+#include "rtos/ThisThread.h"
+
+#include "MathUtils.h"
+
+using namespace leka;
+using namespace std::chrono;
+
+auto SealStrategy::convertToPwmFrom(float angle) const -> float
+{
+	auto res = utils::math::map(angle, kMinAngleInput, kMaxAngleInput, kMinPwmOutput, kMaxPwmOutput);
+	return res;
+}
+
+void SealStrategy::start()
+{
+	auto on_timeout = [this] { stop(); };
+	_timeout.onTimeout(on_timeout);
+	_timeout.start(20s);
+
+	should_stop = false;
+
+	_event_loop.registerCallback([this] { run(); });
+	_event_loop.start();
+}
+
+void SealStrategy::stop()
+{
+	should_stop = true;
+	_event_loop.stop();
+	rtos::ThisThread::sleep_for(100ms);
+	stopMotors();
+	_timeout.stop();
+}
+
+void SealStrategy::run()
+{
+	if (should_stop || _battery.isCharging()) {
+		stop();
+		return;
+	}
+
+	auto angles				 = _imukit.getEulerAngles();
+	auto is_right_tilted	 = angles.roll > 0;
+	auto should_move_forward = angles.pitch > 0;
+
+	auto abs_float = [](float value) { return value > 0 ? value : -value; };
+
+	if (abs_float(angles.roll) > kRollTolerance) {
+		auto speed_offset = convertToPwmFrom(angles.roll > 0 ? angles.roll : -angles.roll);
+		if (is_right_tilted) {
+			if (should_move_forward) {
+				spinRight(kMinPwmOutput, kMinPwmOutput + speed_offset);
+			} else {
+				spinLeft(kMinPwmOutput, kMinPwmOutput + speed_offset);
+			}
+		} else {
+			if (should_move_forward) {
+				spinLeft(kMinPwmOutput + speed_offset, kMinPwmOutput);
+			} else {
+				spinRight(kMinPwmOutput + speed_offset, kMinPwmOutput);
+			}
+		}
```

Comment:
It's been a while since I've coded in C++ !! I kind of remember of us not using ternary operator for a reason but can you put this part in a independant function ?

```
void SealStrategy::performSpin(bool is_right_tilted, bool should_move_forward, float speed_offset) {
    if (is_right_tilted) {
        should_move_forward ? spinRight(kMinPwmOutput, kMinPwmOutput + speed_offset)
                            : spinLeft(kMinPwmOutput, kMinPwmOutput + speed_offset);
    } else {
        should_move_forward ? spinLeft(kMinPwmOutput + speed_offset, kMinPwmOutput)
                            : spinRight(kMinPwmOutput + speed_offset, kMinPwmOutput);
    }
}
```

This would give a nice :

```
if (abs_float(angles.roll) > kRollTolerance) {
    auto speed_offset = convertToPwmFrom(angles.roll > 0 ? angles.roll : -angles.roll);
    performSpin(is_right_tilted, should_move_forward, speed_offset);

} else if (abs_float(angles.pitch) > kPitchTolerance) {
	auto speed = convertToPwmFrom(angles.pitch > 0 ? angles.pitch : -angles.pitch);
	should_move_forward ? moveForward(speed) : moveBackward(speed);

} else {
	stopMotors();

}
```

Reply:
Thank you!

Ternary was mostly not used since it might be hard to read and understand. However I never used them to call a function, only to set variable so I don't know if you can use it as is.

I will let them in if/else statement since it is easier to read and understand what happened to me

I will add your suggestion to move in function these whole if/else statement